src/main/java/rapgod/utils/Parser.java

```diff
@@ -0,0 +1,102 @@
+                    try {
+                        return LocalDateTime.parse(input, dateTimeFormatter3);
+                    } catch (DateTimeParseException e4) {
+                        try {
+                            return LocalDateTime.parse(input, dateTimeFormatter4);
+                        } catch (DateTimeParseException e5) {
+                            throw new IllegalArgumentException("Invalid date or time format. Please use formats like 'dd/MM/yyyy HHmm' or 'MMM dd yyyy'.");
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Parses the string representation of a task and returns the corresponding {@link Task} object.
+     * The string must be in a specific format with the task type indicated (e.g., "D" for Deadline, "E" for Event).
+     *
+     * - Deadline tasks are expected to be in the format: "[D][status] description (by: dueDate)"
+     * - Event tasks are expected to be in the format: "[E][status] description (from: startDate to: endDate)"
+     * - ToDo tasks are expected to be in the format: "[T][status] description"
+     *
+     * The status is indicated by the character at position 8 (e.g., 'X' for completed).
+     *
+     * @param taskString The string representation of the task to be parsed.
+     * @return The corresponding {@link Task} object (either a {@link Deadline}, {@link Event}, or {@link ToDo}).
+     * @throws IllegalArgumentException If the string is null, empty, or does not match the expected format.
+     */
+    public static Task parseToTask(String taskString) {
+        if (taskString == null || taskString.isEmpty()) {
+            throw new IllegalArgumentException("String representation cannot be null or empty");
+        }
+
+        boolean isDone = taskString.charAt(8) == 'X'; // The status is represented by the second character
+        String info = taskString; // The description starts after the brackets
+
+        if (taskString.charAt(4) == 'D') {
+            String deadlineDesc = info.substring(11, info.toLowerCase().indexOf(" (by:"));
+            String due = info.substring(info.toLowerCase().indexOf(" (by:") + 6, info.length() - 1);
+            Deadline deadline = new Deadline(deadlineDesc, due);
+            deadline.setIsDone(isDone);
+            return deadline;
+
+        } else if (taskString.charAt(4) == 'E') {
+            String eventDesc = info.substring(11, info.toLowerCase().indexOf(" (from:"));
+            String from = info.substring(info.toLowerCase().indexOf(" (from:") + 8, info.toLowerCase().indexOf(" to:"));
+            String to = info.substring(info.toLowerCase().indexOf("to:") + 4, info.length() - 1);
+            Event event = new Event(eventDesc, from, to);
+            event.setIsDone(isDone);
+            return event;
+
+        } else {
+            ToDo todo = new ToDo(info.substring(11));
+            todo.setIsDone(isDone);
+            return todo;
+        }
```

Comment:
you may consider using String.spit() to parse parts of the string. This can reduce potential indexing errors :)